File: repos/openjarvis/src/openjarvis/security/audit.py
"""Audit logger — persist security events to SQLite with Merkle hash chain."""

from __future__ import annotations

import hashlib
import json
import sqlite3
from pathlib import Path
from typing import List, Optional, Tuple, Union

from openjarvis.core.config import DEFAULT_CONFIG_DIR
from openjarvis.core.events import Event, EventBus, EventType
from openjarvis.security.types import (
    ScanFinding,
    SecurityEvent,
    SecurityEventType,
    ThreatLevel,
)


class AuditLogger:
    """Append-only SQLite audit log for security events.

    Parameters
    ----------
    db_path:
        Path to the SQLite database file.
    bus:
        Optional event bus — if provided, subscribes to security events
        (``SECURITY_SCAN``, ``SECURITY_ALERT``, ``SECURITY_BLOCK``).
    """

    def __init__(
        self,
        db_path: Union[str, Path] = DEFAULT_CONFIG_DIR / "audit.db",
        bus: Optional[EventBus] = None,
    ) -> None:
        self._db_path = Path(db_path)
        self._db_path.parent.mkdir(parents=True, exist_ok=True)
        self._conn = sqlite3.connect(str(self._db_path))
        self._conn.execute(
            """
            CREATE TABLE IF NOT EXISTS security_events (
                id          INTEGER PRIMARY KEY,
                timestamp   REAL,
                event_type  TEXT,
                findings_json TEXT,
                content_preview TEXT,
                action_taken TEXT,
                row_hash    TEXT DEFAULT '',
                prev_hash   TEXT DEFAULT ''
            )
            """
        )
        self._conn.commit()
        self._migrate_schema()

        if bus is not None:
            bus.subscribe(EventType.SECURITY_SCAN, self._on_event)
            bus.subscribe(EventType.SECURITY_ALERT, self._on_event)
            bus.subscribe(EventType.SECURITY_BLOCK, self._on_event)

    def _migrate_schema(self) -> None:
        """Add row_hash/prev_hash columns if missing (schema migration)."""
        columns = {
            row[1]
            for row in self._conn.execute(
                "PRAGMA table_info(security_events)"
            ).fetchall()
        }
        if "row_hash" not in columns:
            self._conn.execute(
                "ALTER TABLE security_events ADD COLUMN row_hash TEXT DEFAULT ''"
            )
        if "prev_hash" not in columns:
            self._conn.execute(
                "ALTER TABLE security_events ADD COLUMN prev_hash TEXT DEFAULT ''"
            )
        self._conn.commit()

    # -- public API ----------------------------------------------------------

    def log(self, event: SecurityEvent) -> None:
        """Insert a security event into the audit log with Merkle hash chain."""
        findings_json = json.dumps([
            {
                "pattern_name": f.pattern_name,
                "matched_text": f.matched_text,
                "threat_level": f.threat_level.value,
                "start": f.start,
                "end": f.end,
                "description": f.description,
            }
            for f in event.findings
        ])

        # Compute hash chain
        prev_hash = self.tail_hash()
        hash_input = (
            f"{prev_hash}|{event.timestamp}|{event.event_type.value}"
            f"|{findings_json}|{event.content_preview}|{event.action_taken}"
        )
        row_hash = hashlib.sha256(hash_input.encode()).hexdigest()

        self._conn.execute(
            """
            INSERT INTO security_events
                (timestamp, event_type, findings_json, content_preview,
                 action_taken, row_hash, prev_hash)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            """,
            (
                event.timestamp,
                event.event_type.value,
                findings_json,
                event.content_preview,
                event.action_taken,
                row_hash,
                prev_hash,
            ),
        )
        self._conn.commit()

    def query(
        self,
        *,
        event_type: Optional[str] = None,
        since: Optional[float] = None,
        limit: int = 100,
    ) -> List[SecurityEvent]:
        """Query logged security events with optional filters."""
        sql = (
            "SELECT timestamp, event_type, findings_json,"
            " content_preview, action_taken"
            " FROM security_events WHERE 1=1"
        )
        params: list = []

        if event_type is not None:
            sql += " AND event_type = ?"
            params.append(event_type)
        if since is not None:
            sql += " AND timestamp >= ?"
            params.append(since)

        sql += " ORDER BY timestamp DESC LIMIT ?"
        params.append(limit)

        rows = self._conn.execute(sql, params).fetchall()
        events: List[SecurityEvent] = []
        for row in rows:
            ts, etype, findings_json, preview, action = row
            findings_raw = json.loads(findings_json) if findings_json else []
            findings = [
                ScanFinding(
                    pattern_name=f["pattern_name"],
                    matched_text=f["matched_text"],
                    threat_level=ThreatLevel(f["threat_level"]),
                    start=f["start"],
                    end=f["end"],
                    description=f.get("description", ""),
                )
                for f in findings_raw
            ]
            events.append(
                SecurityEvent(
                    event_type=SecurityEventType(etype),
                    timestamp=ts,
                    findings=findings,
                    content_preview=preview or "",
                    action_taken=action or "",
                )
            )
        return events

    def tail_hash(self) -> str:
        """Return the hash of the last row in the chain, or empty string."""
        row = self._conn.execute(
            "SELECT row_hash FROM security_events ORDER BY id DESC LIMIT 1"
        ).fetchone()
        return row[0] if row and row[0] else ""

    def verify_chain(self) -> Tuple[bool, Optional[int]]:
        """Verify the Merkle hash chain integrity.

        Returns
        -------
        tuple
            ``(True, None)`` if the chain is valid, or
            ``(False, row_id)`` where *row_id* is the first broken link.
        """
        rows = self._conn.execute(
            "SELECT id, timestamp, event_type, findings_json,"
            " content_preview, action_taken, row_hash, prev_hash"
            " FROM security_events ORDER BY id"
        ).fetchall()

        expected_prev = ""
        for row in rows:
            rid, ts, etype, fj, preview, action, stored_hash, stored_prev = row
            # Skip rows that predate the Merkle upgrade
            if not stored_hash:
                continue
            # Verify prev_hash link
            if stored_prev != expected_prev:
                return False, rid
            # Verify row_hash
            hash_input = (
                f"{stored_prev}|{ts}|{etype}"
                f"|{fj}|{preview}|{action}"
            )
            computed = hashlib.sha256(hash_input.encode()).hexdigest()
            if computed != stored_hash:
                return False, rid
            expected_prev = stored_hash

        return True, None

    def count(self) -> int:
        """Return the total number of logged security events."""
        row = self._conn.execute(
            "SELECT COUNT(*) FROM security_events"
        ).fetchone()
        return row[0] if row else 0

    def close(self) -> None:
        """Close the SQLite connection."""
        self._conn.close()

    # -- EventBus handler ----------------------------------------------------

    def _on_event(self, event: Event) -> None:
        """Handle an event from the EventBus and log it."""
        data = event.data
        # Map EventType to SecurityEventType
        mapping = {
            EventType.SECURITY_SCAN: SecurityEventType.SECRET_DETECTED,
            EventType.SECURITY_ALERT: SecurityEventType.SECRET_DETECTED,
            EventType.SECURITY_BLOCK: SecurityEventType.SECRET_DETECTED,
        }
        event_type = mapping.get(event.event_type, SecurityEventType.SECRET_DETECTED)

        # Extract findings from event data if present
        findings: List[ScanFinding] = []
        for f in data.get("findings", []):
            findings.append(
                ScanFinding(
                    pattern_name=f.get("pattern", ""),
                    matched_text="",
                    threat_level=ThreatLevel(f.get("threat", "low")),
                    start=0,
                    end=0,
                    description=f.get("description", ""),
                )
            )

        sec_event = SecurityEvent(
            event_type=event_type,
            timestamp=event.timestamp,
            findings=findings,
            content_preview=data.get("content_preview", ""),
            action_taken=data.get("mode", ""),
        )
        self.log(sec_event)


__all__ = ["AuditLogger"]

File: repos/openjarvis/rust/crates/openjarvis-security/src/audit.rs
//! Audit logger — persist security events to SQLite with Merkle hash chain.

use crate::types::{ScanFinding, SecurityEvent, SecurityEventType};
use openjarvis_core::OpenJarvisError;
use rusqlite::Connection;
use sha2::{Digest, Sha256};
use std::path::{Path, PathBuf};

pub struct AuditLogger {
    conn: Connection,
    _db_path: PathBuf,
}

impl AuditLogger {
    pub fn new(db_path: &Path) -> Result<Self, OpenJarvisError> {
        if let Some(parent) = db_path.parent() {
            std::fs::create_dir_all(parent).map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(e))
            })?;
        }

        let conn = Connection::open(db_path).map_err(|e| {
            OpenJarvisError::Io(std::io::Error::other(
                e.to_string(),
            ))
        })?;

        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS security_events (
                id          INTEGER PRIMARY KEY,
                timestamp   REAL,
                event_type  TEXT,
                findings_json TEXT,
                content_preview TEXT,
                action_taken TEXT,
                row_hash    TEXT DEFAULT '',
                prev_hash   TEXT DEFAULT ''
            )",
        )
        .map_err(|e| {
            OpenJarvisError::Io(std::io::Error::other(
                e.to_string(),
            ))
        })?;

        Ok(Self {
            conn,
            _db_path: db_path.to_path_buf(),
        })
    }

    pub fn log(&self, event: &SecurityEvent) -> Result<(), OpenJarvisError> {
        let findings_json = serde_json::to_string(&event.findings).unwrap_or_default();
        let prev_hash = self.tail_hash();

        let hash_input = format!(
            "{}|{}|{:?}|{}|{}|{}",
            prev_hash,
            event.timestamp,
            event.event_type,
            findings_json,
            event.content_preview,
            event.action_taken
        );
        let row_hash = hex_sha256(&hash_input);

        self.conn
            .execute(
                "INSERT INTO security_events
                    (timestamp, event_type, findings_json, content_preview,
                     action_taken, row_hash, prev_hash)
                VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)",
                rusqlite::params![
                    event.timestamp,
                    format!("{:?}", event.event_type),
                    findings_json,
                    event.content_preview,
                    event.action_taken,
                    row_hash,
                    prev_hash,
                ],
            )
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;

        Ok(())
    }

    pub fn tail_hash(&self) -> String {
        self.conn
            .query_row(
                "SELECT row_hash FROM security_events ORDER BY id DESC LIMIT 1",
                [],
                |row| row.get::<_, String>(0),
            )
            .unwrap_or_default()
    }

    /// Verify the Merkle hash chain integrity.
    /// Returns `(true, None)` if valid, or `(false, Some(row_id))` for first broken link.
    pub fn verify_chain(&self) -> Result<(bool, Option<i64>), OpenJarvisError> {
        let mut stmt = self
            .conn
            .prepare(
                "SELECT id, timestamp, event_type, findings_json,
                        content_preview, action_taken, row_hash, prev_hash
                 FROM security_events ORDER BY id",
            )
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;

        let rows = stmt
            .query_map([], |row| {
                Ok((
                    row.get::<_, i64>(0)?,
                    row.get::<_, f64>(1)?,
                    row.get::<_, String>(2)?,
                    row.get::<_, String>(3)?,
                    row.get::<_, Option<String>>(4)?,
                    row.get::<_, Option<String>>(5)?,
                    row.get::<_, String>(6)?,
                    row.get::<_, String>(7)?,
                ))
            })
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;

        let mut expected_prev = String::new();

        for row_result in rows {
            let (rid, ts, etype, fj, preview, action, stored_hash, stored_prev) =
                row_result.map_err(|e| {
                    OpenJarvisError::Io(std::io::Error::other(
                        e.to_string(),
                    ))
                })?;

            if stored_hash.is_empty() {
                continue;
            }

            if stored_prev != expected_prev {
                return Ok((false, Some(rid)));
            }

            let hash_input = format!(
                "{}|{}|{}|{}|{}|{}",
                stored_prev,
                ts,
                etype,
                fj,
                preview.unwrap_or_default(),
                action.unwrap_or_default()
            );
            let computed = hex_sha256(&hash_input);
            if computed != stored_hash {
                return Ok((false, Some(rid)));
            }
            expected_prev = stored_hash;
        }

        Ok((true, None))
    }

    pub fn count(&self) -> i64 {
        self.conn
            .query_row("SELECT COUNT(*) FROM security_events", [], |row| {
                row.get(0)
            })
            .unwrap_or(0)
    }

    pub fn query(
        &self,
        event_type: Option<&str>,
        since: Option<f64>,
        limit: usize,
    ) -> Result<Vec<SecurityEvent>, OpenJarvisError> {
        let mut sql = String::from(
            "SELECT timestamp, event_type, findings_json, content_preview, action_taken
             FROM security_events WHERE 1=1",
        );
        let mut params: Vec<Box<dyn rusqlite::types::ToSql>> = Vec::new();

        if let Some(et) = event_type {
            sql.push_str(" AND event_type = ?");
            params.push(Box::new(et.to_string()));
        }
        if let Some(s) = since {
            sql.push_str(" AND timestamp >= ?");
            params.push(Box::new(s));
        }
        sql.push_str(" ORDER BY timestamp DESC LIMIT ?");
        params.push(Box::new(limit as i64));

        let param_refs: Vec<&dyn rusqlite::types::ToSql> =
            params.iter().map(|p| p.as_ref()).collect();

        let mut stmt = self.conn.prepare(&sql).map_err(|e| {
            OpenJarvisError::Io(std::io::Error::other(
                e.to_string(),
            ))
        })?;

        let rows = stmt
            .query_map(param_refs.as_slice(), |row| {
                Ok((
                    row.get::<_, f64>(0)?,
                    row.get::<_, String>(1)?,
                    row.get::<_, Option<String>>(2)?,
                    row.get::<_, Option<String>>(3)?,
                    row.get::<_, Option<String>>(4)?,
                ))
            })
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;

        let mut events = Vec::new();
        for row_result in rows {
            let (ts, _etype, findings_json, preview, action) =
                row_result.map_err(|e| {
                    OpenJarvisError::Io(std::io::Error::other(
                        e.to_string(),
                    ))
                })?;

            let findings: Vec<ScanFinding> = findings_json
                .as_deref()
                .and_then(|s| serde_json::from_str(s).ok())
                .unwrap_or_default();

            events.push(SecurityEvent {
                event_type: SecurityEventType::SecretDetected,
                timestamp: ts,
                findings,
                content_preview: preview.unwrap_or_default(),
                action_taken: action.unwrap_or_default(),
            });
        }

        Ok(events)
    }
}

fn hex_sha256(input: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(input.as_bytes());
    format!("{:x}", hasher.finalize())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::ThreatLevel;

    #[test]
    fn test_audit_log_and_verify() {
        let dir = tempfile::tempdir().unwrap();
        let db_path = dir.path().join("test_audit.db");
        let logger = AuditLogger::new(&db_path).unwrap();

        let event = SecurityEvent {
            event_type: SecurityEventType::SecretDetected,
            timestamp: 1000.0,
            findings: vec![ScanFinding {
                pattern_name: "openai_key".into(),
                matched_text: "sk-test123".into(),
                threat_level: ThreatLevel::Critical,
                start: 0,
                end: 10,
                description: "OpenAI API key".into(),
            }],
            content_preview: "test".into(),
            action_taken: "warn".into(),
        };
        logger.log(&event).unwrap();

        assert_eq!(logger.count(), 1);
        let (valid, _) = logger.verify_chain().unwrap();
        assert!(valid);
    }

    #[test]
    fn test_audit_chain_integrity() {
        let dir = tempfile::tempdir().unwrap();
        let db_path = dir.path().join("test_chain.db");
        let logger = AuditLogger::new(&db_path).unwrap();

        for i in 0..5 {
            let event = SecurityEvent {
                event_type: SecurityEventType::SecretDetected,
                timestamp: 1000.0 + i as f64,
                findings: vec![],
                content_preview: format!("event {}", i),
                action_taken: "warn".into(),
            };
            logger.log(&event).unwrap();
        }

        assert_eq!(logger.count(), 5);
        let (valid, _) = logger.verify_chain().unwrap();
        assert!(valid);
    }
}
